internal/ansible/ansible.go

```diff
@@ -0,0 +1,221 @@
+	}
+	if err := os.MkdirAll(filepath.Join(privateDataDir, "artifacts", id, "job_events"), 0755); err != nil {
+		return nil, fmt.Errorf("cannot create private state dir: %v", err)
+	}
+
+	// precreate the status file so that we can watch for when it gets written
+	// to.
+	status, err := os.Create(filepath.Join(privateDataDir, "artifacts", id, "status"))
+	if err != nil {
+		return nil, fmt.Errorf("cannot create status file: %v", err)
+	}
+	_ = status.Close()
+
+	events := make(chan Event)
+
+	// publish an "executor_on_start" event to signal cloud connector that a run
+	// event has started. This is run on a goroutine in case the events
+	// channel doesn't have a receiver connected yet to avoid blocking the
+	// continuation of this function.
+	go func() {
+		events <- Event{
+			Event:     "executor_on_start",
+			UUID:      uuid.New().String(),
+			Counter:   -1,
+			Stdout:    "",
+			StartLine: 0,
+			EndLine:   0,
+			EventData: struct {
+				CRCDispatcherCorrelationID string `json:"crc_dispatcher_correlation_id"`
+				CRCDispatcherErrorCode     string `json:"crc_dispatcher_error_code"`
+				CRCDispatcherErrorDetails  string `json:"crc_dispatcher_error_details"`
+			}{
+				CRCDispatcherCorrelationID: correlationID,
+			},
+		}
+	}()
+
+	// started is a closure that's passed to the StartProcess function as its
+	// exec.ProcessStartedFunc handler. This function is invoked on a goroutine
+	// after the process has been started. It exists as a named closure within
+	// this function to capture some relevant values that aren't otherwise
+	// passed in as part of the exec.ProcessStartedFunc signature (id,
+	// privateDataDir, etc.).
+	started := func(pid int, stdout, stderr io.ReadCloser) {
+		log.Debugf("process started: pid=%v runner_ident=%v", pid, id)
+
+		// start a goroutine to watch for event files being written to the
+		// job_events directory. When a relevant event file is detected, it gets
+		// marshaled into JSON and sent to the events channel.
+		go func() {
+			c := make(chan notify.EventInfo, 1)
+			if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "job_events"), c, notify.InMovedTo); err != nil {
+				log.Errorf("failed to watch private data dir: dir=%v err=%v", privateDataDir, err)
+				return
+			}
+			defer notify.Stop(c)
+
+			for e := range c {
+				log.Debugf("notify event: event=%v path=%v", e.Event(), e.Path())
+				switch e.Event() {
+				case notify.InMovedTo:
+					if strings.HasSuffix(e.Path(), ".json") {
+						data, err := os.ReadFile(e.Path())
+						if err != nil {
+							log.Errorf("cannot read file: file=%v error=%v", e.Path(), err)
+							continue
+						}
+
+						var event Event
+						if err := json.Unmarshal(data, &event); err != nil {
+							log.Errorf("cannot unmarshal data: data=%v error=%v", data, err)
+							continue
+						}
+
+						events <- event
+						log.Infof("event sent: event=%v", event)
+					}
+				}
+			}
+		}()
```

Comment:
Is this go routine actually terminated, when related process is finished? The `started` closure is terminated, but I'm not sure about this go routine.

Reply:
My suspicion for this case was also correct. When you create some `foo.json` file with some foo content. Something like this:

```json
{"foo": "bar"}
```

And you move this file to `job_events` directory of finished job:

```
mv ~/foo.json /var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/job_events/foo.json
```

Then it is obvious that the go routine is still running, but it terminates whole worker as you can see:

```
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: 2024/08/15 11:27:20 notify event: event=notify.InMovedTo path=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/job_events/foo.json
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: 2024/08/15 11:27:20 >>> terminating goroutine for watching data dir: dir=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: panic: send on closed channel
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: goroutine 14 [running]:
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: github.com/redhatinsights/rhc-worker-playbook/internal/ansible.RunPlaybook.func2.1()
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]:         /home/jhnidek/github/redhat-insights/rhc-worker-playbook/internal/ansible/ansible.go:119 +0x60b
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: created by github.com/redhatinsights/rhc-worker-playbook/internal/ansible.RunPlaybook.func2 in goroutine 11
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]:         /home/jhnidek/github/redhat-insights/rhc-worker-playbook/internal/ansible/ansible.go:93 +0x1bb
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Failed with result 'exit-code'.
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Consumed 1.934s CPU time.
```

... Yeah, I added some extra debug print to the code.